examples/ann/main.go

```diff
@@ -0,0 +1,252 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Bram Gruneir (bram+code@cockroachlabs.com)
+
+package main
+
+import (
+	"bytes"
+	"database/sql"
+	"flag"
+	"fmt"
+	"math"
+	"math/rand"
+	"time"
+
+	"github.com/cockroachdb/cockroach/security"
+	"github.com/cockroachdb/cockroach/security/securitytest"
+	"github.com/cockroachdb/cockroach/server"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+var inputLayerNodes = flag.Int("input-layer-nodes", 32, "Number of nodes in the input layer.")
+var hiddenLayerNodes = flag.Int("hidden-layer-nodes", 9, "Number of nodes in the hidden layer.")
+var outputLayerNodes = flag.Int("output-layer-nodes", 32, "Number of nodes in the hidden layer.")
+var hiddenLayers = flag.Int("hidden-layer-count", 1, "Number of hidden layers.")
+var totalIterations = flag.Int("iterations", 100, "Number of iterations.")
+
+var rng *rand.Rand
+var maxInput int64
+var testServer *server.TestServer
+
+func createConnection(database bool) *sql.DB {
+	url := fmt.Sprintf("https://root@%s?certs=test_certs", testServer.ServingAddr())
+	if database {
+		url += "&database=ann"
+	}
+	var db *sql.DB
+	var err error
+	if db, err = sql.Open("cockroach", url); err != nil {
+		log.Fatal(err)
+	}
+	return db
+}
+
+// getNodeID returns a node id based on it's level and index into its layer.
+// This way, we can store all nodes in a single table on the db.
+func getNodeID(layer, num int) int {
+	return (layer * 100000) + num
+}
+
+func getLayerNodeCount(layer int) int {
+	if layer == 0 {
+		return *inputLayerNodes
+	}
+	if layer == 1+*hiddenLayers {
+		return *outputLayerNodes
+	}
+	return *hiddenLayerNodes
+}
+
+func addInput() int64 {
+	newInput := rng.Int63n(maxInput)
+	tempInput := newInput
+	var binary bytes.Buffer
+	db := createConnection(true)
+	defer db.Close()
+	tx, err := db.Begin()
+	if err != nil {
+		log.Fatal(err)
+	}
+	for i := 0; i < *inputLayerNodes; i++ {
+		bit := tempInput & 1
+		tempInput = tempInput >> 1
+		if bit == 0 {
+			binary.WriteString("0")
+		} else {
+			binary.WriteString("1")
+		}
+		if _, err := tx.Exec(`UPDATE nodes SET value=$1 WHERE id=$2`, float64(bit), getNodeID(0, i)); err != nil {
+			log.Fatal(err)
+		}
+		// TODO(bram): Remove this once we have the feed forward network step working. This is only temporary to test
+		// parsing the output nodes.
+		if _, err := tx.Exec(`UPDATE nodes SET value=$1 WHERE id=$2`, float64(bit), getNodeID(*hiddenLayers+1, i)); err != nil {
+			log.Fatal(err)
+		}
+	}
+	if err = tx.Commit(); err != nil {
+		log.Fatal(err)
+	}
+	fmt.Printf("Input: %d or %s\n", newInput, binary.String())
+	return newInput
+}
+
+func getOutput() int64 {
+	var result int64
+	var binary bytes.Buffer
+	db := createConnection(true)
```

Comment:
why are you creating so many connections? `sql.DB` is thread-safe, and will use multiple connections under the hood.